src/Shinkansen.cpp

```diff
@@ -0,0 +1,159 @@
+#include <stdio.h>
+#include <sys/socket.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <netinet/in.h>
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <sstream>
+
+#include "Shinkansen.h"
+#include "Utils.h"
+#include "Request.h"
+#include "Response.h"
+#include "Constants.h"
+
+using namespace std;
+
+Shinkansen::Shinkansen(int port)
+{
+  this->port = port;
+  this->address = this->AddressConfig();
+  this->Initialize();
+}
+
+void Shinkansen::Initialize()
+{
+  // Creating socket file descriptor
+  if ((this->serverSocket = socket(AF_INET, SOCK_STREAM, 0)) == 0)
+  {
+    perror("In socket");
+    exit(EXIT_FAILURE);
+  }
+
+  if (bind(this->serverSocket, (struct sockaddr *)&address, sizeof(this->address)) < 0)
+  {
+    perror("In bind");
+    exit(EXIT_FAILURE);
+  }
+  if (listen(this->serverSocket, 10) < 0)
+  {
+    perror("In listen");
+    exit(EXIT_FAILURE);
+  }
+
+  cout << "Server listening on port " << this->port << "..." << endl;
+  this->Listen();
+}
+
+void Shinkansen::Listen()
+{
+
+  while (1)
+  {
+    int addrlen = sizeof(this->address);
+
+    if ((this->clientSocket = accept(this->serverSocket, (struct sockaddr *)&address, (socklen_t *)&addrlen)) < 0)
+    {
+      perror("In accept");
+      exit(EXIT_FAILURE);
+    }
+
+    // form request object
+    Request request = this->ParseRequest();
+
+    this->SendFile(request);
+  }
+}
+
+void Shinkansen::SendFile(Request req)
+{
+  Response *res = new Response();
+  ifstream in;
+  pair<int, string> statusCode;
+  stringstream body;
+
+  // return homepage
+  if (req.GetRoute() == "/" || req.GetRoute() == "")
+  {
+    in.open("public/index.html", ios::binary);
+    if (in.is_open())
+    {
+      statusCode.first = 200;
+      statusCode.second = "Ok";
+      res->SetStatusCode(statusCode);
+      res->SetContentType("text/html");
+
+      // read data and assign to body
+      body << in.rdbuf();
+      res->SetBody(body.str());
+
+      this->SendRaw(*res);
+      delete res;
+      return; 
+    }
+  }
+  // try to find file
+  in.open("public" + req.GetRoute(), ios::binary);
+
+  if (!in.is_open())
+  {
+
+    statusCode.first = 404;
+    statusCode.second = "Not Found";
+    res->SetStatusCode(statusCode);
+    res->SetContentType("text/html");
+    res->SetBody("<h1>404 Not Found</h1>\n<p>Please check your request url!</p>");
+    this->SendRaw(*res);
+    delete res;
+    return;
+  }
+  // response config
+
+  statusCode.first = 200;
+  statusCode.second = "Ok";
+  res->SetStatusCode(statusCode);
+  res->SetContentType(Constants::GetContentTypeFromExt(req.GetFileExtension()));
+
+  // read data and assign to body
+  body << in.rdbuf();
+  res->SetBody(body.str());
+
+  this->SendRaw(*res);
+  delete res;
+}
+
+void Shinkansen::SendRaw(Response &res)
+{
+  string s = res.ToString();
+  char rawResponse[s.length()];
```

Comment:
This is a VLA which is not legal C++ (but a common extension) ... it is not standard because it is not safe and there are usually better options ... in fact this one will probably blow up your stack if the response is big enough (try with a file > ~8MB which is the usual linux stack size)